tests/core/arch-x86/stalker-x86.c

```diff
@@ -375,6 +401,239 @@ invoke_flat (TestStalkerFixture * fixture,
+  g_clear_object (&fixture->transformer);
+  fixture->transformer = gum_stalker_transformer_make_from_callback (
+      prefetch_transformer, NULL, NULL);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+  gum_stalker_deactivate (fixture->stalker);
+
+  /* Run the child */
+  prefetch_run_child (fixture->stalker, &runner_range,
+      compile_pipes[STDOUT_FILENO], execute_pipes[STDOUT_FILENO]);
+
+  /* Read the results */
+  compiled_run1 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (compile_pipes[STDIN_FILENO], compiled_run1);
+  executed_run1 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (execute_pipes[STDIN_FILENO], executed_run1);
+
+  compiled_size_run1 = g_hash_table_size (compiled_run1);
+  g_print ("\tcompiled: %d\n", compiled_size_run1);
+
+  executed_size_run1 = g_hash_table_size (executed_run1);
+  g_print ("\texecuted: %d\n", executed_size_run1);
+
+  g_assert_cmpuint (compiled_size_run1, >, 0);
+  g_assert_cmpuint (compiled_size_run1, ==, executed_size_run1);
+
+  /* Prefetch the blocks */
+  gum_stalker_set_trust_threshold (fixture->stalker, 0);
+  g_hash_table_iter_init (&iter, compiled_run1);
+  while (g_hash_table_iter_next (&iter, &iter_key, &iter_value))
+  {
+    for (i = 0; i < trust; i++)
+      gum_stalker_prefetch (fixture->stalker, iter_key);
+  }
+  gum_stalker_set_trust_threshold (fixture->stalker, trust);
+
+  /* Run the child again */
+  prefetch_run_child (fixture->stalker, &runner_range,
+      compile_pipes[STDOUT_FILENO], execute_pipes[STDOUT_FILENO]);
+
+  /* Read the results */
+  compiled_run2 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (compile_pipes[STDIN_FILENO], compiled_run2);
+  executed_run2 = g_hash_table_new (NULL, NULL);
+  prefetch_read_blocks (execute_pipes[STDIN_FILENO], executed_run2);
+
+  compiled_size_run2 = g_hash_table_size (compiled_run2);
+  g_print ("\tcompiled2: %d\n", compiled_size_run2);
+
+  executed_size_run2 = g_hash_table_size (executed_run2);
+  g_print ("\texecuted2: %d\n", executed_size_run2);
+
+  g_assert_cmpuint (compiled_size_run2, ==, 0);
+  g_assert_cmpuint (executed_size_run2, ==, executed_size_run1);
+
+
+  /* Free resources */
+  g_hash_table_destroy (compiled_run2);
+  g_hash_table_destroy (executed_run2);
+  g_hash_table_destroy (compiled_run1);
+  g_hash_table_destroy (executed_run1);
+
+  gum_stalker_unfollow_me (fixture->stalker);
+
+  g_print ("DONE\n");
+
+}
+
+static void
+prefetch_transformer (GumStalkerIterator * iterator,
+                   GumStalkerOutput * output,
+                   gpointer user_data)
+{
+  const cs_insn * instr;
+  gboolean begin = TRUE;
+  gpointer block_address;
+  gpointer * callout_ctx;
+
+  while (gum_stalker_iterator_next (iterator, &instr))
+  {
+    if (begin)
+    {
+      block_address = GSIZE_TO_POINTER (instr->address);
+      callout_ctx = gum_malloc (sizeof (gpointer));
+      *callout_ctx = block_address;
```

Comment:
(But why the indirection if it's only one pointer?)

Reply:
I'll take a look at this one.